### ISSUE TITLE
Hooks not working in version 2.0.1+cu118

### ISSUE THREAD

**Issue by Khisrow1** [pytorch/pytorch#102374]
### 🐛 Describe the bug

When writing forward hooks on a layer of a model, I realized they are not being called at all.

```python
# inside __init__ function
for name, module in self.model.named_modules():
    if attention_layer_name in name:
    print(f"Module hook registered for: {name}")
    module.register_forward_hook(lambda m, i, o: attentions.append(o))
```
Expected Behavior: When I forward an example through the model, the hooks should be called after specifically hooked layers are done forwarding the weights

I did try it on torch version 1.13.1, and it seems to work fine there.

### Versions

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.1+cu118
[pip3] torchaudio==2.0.2+cu118
[pip3] torchvision==0.15.2+cu118
[pip3] triton==2.0.0
[conda] Could not collect

**Comment by mikaylagawarecki** [comment]
Hey, could you provide a minimally reproducible example please?

I tried the following script with a random layer name

```
import torch
import torch.nn as nn
import torch.nn.functional as F

class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = torch.flatten(x, 1) # flatten all dimensions except batch
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x


net = Net()
linears = []
for name, module in net.named_modules():
  print(name)
  if "fc" in name:
    print(f"Module hook registered for: {name}")
    module.register_forward_hook(lambda m, i, o: linears.append(o))


inp = torch.randn((1024, 3, 32, 32))
net(inp)
print(linears)
```

and the list appended to in the forward hook was accurately populated

**Comment by Khisrow1** [comment]
So, I did run your snippet, and it seems to work for your example. In my case, however, I was using it with a pretrained transformer model.
```python
import torch

model = torch.hub.load('facebookresearch/deit:main',  'deit_tiny_patch16_224', pretrained=True)
model.eval()

linears = []
for name, module in model.named_modules():
  if "attn_drop" in name:
    print(f"Module hook registered for: {name}")
    module.register_forward_hook(lambda m, i, o: linears.append(o))


print(linears)
```
Output:
```
Module hook registered for: blocks.0.attn.attn_drop
Module hook registered for: blocks.1.attn.attn_drop
Module hook registered for: blocks.2.attn.attn_drop
Module hook registered for: blocks.3.attn.attn_drop
Module hook registered for: blocks.4.attn.attn_drop
Module hook registered for: blocks.5.attn.attn_drop
Module hook registered for: blocks.6.attn.attn_drop
Module hook registered for: blocks.7.attn.attn_drop
Module hook registered for: blocks.8.attn.attn_drop
Module hook registered for: blocks.9.attn.attn_drop
Module hook registered for: blocks.10.attn.attn_drop
Module hook registered for: blocks.11.attn.attn_drop
[]
```
So, it kind of works, but not for this pretrained model. But then the whole point of the hook is to use it in circumstances like this.

**Comment by albanD** [comment]
You don't actually evaluate the module in your example. So it is expected that the hooks are never triggered right?

**Comment by Khisrow1** [comment]
I'm not 100% percent sure on how you mean "evaluate", perhaps you could clarify a bit :)
As you can see in the example, the hooks are supposed to populate the list with their outputs, which it does not. As mentioned before, the same code works as expected with the previous version.

**Comment by Khisrow1** [comment]
I think I forgot to write it here, but I did evaluate the model. As can be seen here:

``` python
import torch

model = torch.hub.load('facebookresearch/deit:main',  'deit_tiny_patch16_224', pretrained=True)
model.eval()

linears = []
for name, module in model.named_modules():
  if "attn_drop" in name:
    print(f"Module hook registered for: {name}")
    module.register_forward_hook(lambda m, i, o: linears.append(o))


ten = torch.randn((1, 3, 224, 224))
out = model(ten)

print(linears)
```
I even ran the thing on colab right now: https://colab.research.google.com/drive/12biWdbUBxWvonb3p3JWbJ1fbrCjnL_X7?usp=sharing

**Comment by albanD** [comment]
Thanks for the full example. This works just fine when using a regular Module. So my guess is that the timm model in question is doing something shady.

**Comment by jbschlosser** [comment]
@Khisrow1 Does the issue persist with a nightly build of PyTorch? I can't repro it with your snippet on a fairly new build, so perhaps it has been fixed already.

**Comment by tcapelle** [comment]
Just found this today on Pytorch 2.0.1:
<img width="645" alt="image" src="https://github.com/pytorch/pytorch/assets/18441985/3bc37d46-fbd2-4788-8c6b-74808eb97f6e">

The hook is only called if you call the module with the `__call__` method and not when using `forward`, is this a bug?

**Comment by mikaylagawarecki** [comment]
@tcapelle This is expected behavior. Hooks are run in the `__call__` function, so if you call `.forward` and have hooks in your model, the hooks won’t have any effect

**Comment by tcapelle** [comment]
Lol, as the function is called `register_forward_hook` and don't work when calling `forward`

**Comment by kevin-ssy** [comment]
Same issue here, any solutions?

**Comment by carlz1020** [comment]
I fixed this by adding this:
```
import os
os.environ['TIMM_FUSED_ATTN'] = '0'
import timm
....
```
Same issue to me when using timm. 
check timm codes on attention blocks:
 `self.fused_attn = use_fused_attn()` used for calling `F.scaled_dot_product_attention` above pytorch 2.0
and in the forward part, and skip original `self.attn_drop = nn.Dropout(attn_drop)`.
I think thus why no inputs and outputs when hooking `attn_drop`.
